### ISSUE TITLE
Is there a way to configure multiple stand-alone instances of azure signal r service in multi-tenant application

### ISSUE THREAD

**Issue by asadmalik3** [Azure/azure-signalr#1561]
### Is your feature request related to a problem? Please describe.
I have a multi-tenant application where users can configure their own azure signal r service instance. During negotiation, on the basis of domain or some other parameter in the query string, I should return user details of customer's own endpoint.

### Describe the solution you'd like
Is there a way to achieve this? This comes close to design of factory pattern.

**Comment by vicancy** [comment]
How do your customers configure their SignalR instances? Does your app server have the mapping between the user and the SignalR endpoint?

**Comment by asadmalik3** [comment]
@vicancy Yes, we let users configure signal r endpoint through web app and store this end point in database for every customer separately.

**Comment by vicancy** [comment]
Do you have Hub logic in your app server? Do you know the list of SignalR endpoints that your customer can select beforehand?

**Comment by asadmalik3** [comment]
@vicancy Yes, hub logic is in my app server. SignalR endpoints for customers are stored in the database and I can access those endpoints.

**Comment by vicancy** [comment]
Can these endpoints change when new customers come in? 

If not, you can try this multi-endpoint approach https://docs.microsoft.com/en-us/azure/azure-signalr/signalr-howto-scale-multi-instances#how-to-add-multiple-endpoints-from-code, and use this custom router logic https://docs.microsoft.com/en-us/azure/azure-signalr/signalr-howto-scale-multi-instances#customize-routing-algorithm to route different customers to different endpoints with your own `GetNegotiateEndpoint` logic.

**Comment by asadmalik3** [comment]
@vicancy Endpoints can definitely change, customers can add/remove endpoints from web application. 

I tried to write a custom routing algorithm, but it requires me to declare endpoints at the startup. My issue is that endpoints can be added/removed/changed by the user.

**Comment by vicancy** [comment]
The SDK actually did a lot of work to support dynamic endpoint hot-reload from configurations. The work is mainly to sync between multiple app-server instances to guarantee every app-server handles the same list of endpoints. So as this doc https://github.com/Azure/azure-signalr/blob/dev/docs/sharding.md#dynamic-scale-serviceendpoints points out, if you want to support dynamic endpoints, you can add your custom configuration provider as similar to this https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-3.1#custom-configuration-provider-1 supporting reloading changes from your database.

**Comment by asadmalik3** [comment]
@vicancy This seems possible on paper, I will try it out and will update you. Thanks.
